querybook/webapp/lib/python/python-worker.ts

```diff
@@ -0,0 +1,219 @@
+import { expose } from 'comlink';
+import { loadPyodide, PyodideInterface, version } from 'pyodide';
+
+import type { PyProxy } from 'pyodide/ffi';
+
+import { patchPyodide } from './patch';
+import { PythonKernel } from './types';
+
+// Constants
+const INDEX_URL = `https://cdn.jsdelivr.net/pyodide/v${version}/full/`;
+const DEFAULT_PACKAGES = ['micropip', 'numpy', 'pandas', 'matplotlib'];
+
+/**
+ * PyodideKernel - Handles Pyodide initialization and execution
+ * Implements the PythonKernel interface
+ */
+class PyodideKernel implements PythonKernel {
+    public version = version;
+
+    private loadPyodidePromise: Promise<void> | null = null;
+    private namespaces: Record<number, PyProxy> = {};
+    private executionCountByNS: Record<number, number> = {};
+    private pyodide: PyodideInterface | null = null;
+    private interruptBuffer: Uint8Array | null = null;
+
+    /**
+     * Initialize pyodide with specified packages
+     *
+     * @param packages - Additional packages to load via micropip
+     */
+    public async initialize(packages: string[] = []): Promise<void> {
+        if (this.loadPyodidePromise !== null) {
+            throw new Error('Pyodide was already initialized');
+        }
+
+        this.loadPyodidePromise = this._loadPyodide(packages);
+        return this.loadPyodidePromise;
+    }
+
+    /**
+     * Run Python code in the specified namespace
+     *
+     * @returns The result of the execution, if any
+     */
+    public async runPython(
+        code: string,
+        namespaceId?: number,
+        stdoutCallback?: ((text: string) => void) | null,
+        stderrCallback?: ((text: string) => void) | null
+    ): Promise<void> {
+        if (!this.loadPyodidePromise) {
+            throw new Error(
+                'Pyodide not initialized. Call initialize() first.'
+            );
+        }
+
+        await this.loadPyodidePromise;
+
+        if (!this.pyodide || !this.interruptBuffer) {
+            throw new Error('Pyodide initialization failed');
+        }
+
+        // Reset interrupt buffer and set status
+        this.interruptBuffer[0] = 0;
+        this.executionCountByNS[namespaceId] =
```

Comment:
namespaceid could be undefined, which is kind of confusing to set in a map, could we put some default name space then?

Reply:
added undefined check. will create a temporary namespace for any adhoc run without namespace id. dont want to persist any variables using a default namespace.